aiida_mlip/data/model.py

```diff
@@ -0,0 +1,159 @@
+"""Define Model Data type in AiiDA"""
+
+import hashlib
+from pathlib import Path
+from urllib import request
+from urllib.parse import urlparse
+
+from aiida.orm import SinglefileData
+
+
+def _calculate_hash(file: str | Path) -> str:
+    """Function to calculate file hashes
+    param: file: path to the file you want to calculate the hash for (str)
+    """
+    # Calculate hash
+    buf_size = 65536  # reading 64kB (arbitrary) at a time
+    sha256 = hashlib.sha256()
+    with open(file, "rb") as f:
+        # calculating sha in chunks rather than 1 large pass
+        while data := f.read(buf_size):
+            sha256.update(data)
+    file_hash = sha256.hexdigest()
+    return file_hash
+
+
+def _check_existing_file(
+    arch_path: str | Path, file: str | Path, architecture: str, cache_path: str | Path
+) -> tuple[Path, str]:
+    """Function to check if a file already exists and return a file path and architecture
+    param arch_path: Path to the folder in which the file is stored (str or pathlib.Path).
+    param file: Path to the file (str or pathlib.Path)
+    param architecture: mlip architecture of the model
+    param cache_path: Path to the parent folder of arch_path (str or pathlib.Path)
+    return: A tuple of a PAth and a str containing the path of the file of interest and the architecture
+    """
+    file_hash = _calculate_hash(file)
+    for existing_file in (
+        ex_file
+        for ex_file in arch_path.rglob("*")
+        if ex_file.is_file() and not ex_file.samefile(file)
+    ):
```

Comment:
I think this may be cleaner as a filter
```suggestion
    def is_diff_file(curr_path: Path):
        return curr_path.is_file() and not curr_path.samefile(path)

    for existing_file in filter(is_diff_file, arch_path.rglob("*")):
```